base/lock.jl

```diff
@@ -342,28 +346,44 @@ end
 
 
 """
-    Event()
+    Event([autoreset=false])
 
 Create a level-triggered event source. Tasks that call [`wait`](@ref) on an
 `Event` are suspended and queued until [`notify`](@ref) is called on the `Event`.
 After `notify` is called, the `Event` remains in a signaled state and
-tasks will no longer block when waiting for it.
+tasks will no longer block when waiting for it, until `reset` is called.
+
+If `autoreset` is true, at most one thread will be released from `wait` for
+each call to `notify`.
+
+This provides an acquire & release memory ordering on notify/wait.
 
 !!! compat "Julia 1.1"
     This functionality requires at least Julia 1.1.
+
+!!! compat "Julia 1.8"
+    The `autoreset` functionality and memory ordering guarantee requires at least Julia 1.8.
 """
 mutable struct Event
     notify::Threads.Condition
-    set::Bool
-    Event() = new(Threads.Condition(), false)
+    autoreset::Bool
+    @atomic set::Bool
+    Event(autoreset::Bool=false) = new(Threads.Condition(), autoreset, false)
 end
 
 function wait(e::Event)
-    e.set && return
+    if e.autoreset
+        (@atomicswap :acquire_release e.set = false) && return
+    else
+        (@atomic e.set) && return # full barrier also
+    end
     lock(e.notify)
     try
-        while !e.set
+        if !e.set
             wait(e.notify)
+            @atomic e.set # full barrier also
+        elseif e.autoreset
+            @atomic e.set = false
```

Comment:
I think this needs to be an atomic swap, and then possibly go around the `while` loop again, in case another thread reaches `wait` just before this line?

Reply:
Aha, or I guess in that case `e.set` won't have been set, so it's ok?

Reply:
There's a lock here, so that situation can and must never occur

Reply:
ooh, I see it might get cleared outside this loop, okay